### DISCUSSION TITLE
Using Modal routes with Remix

### DISCUSSION THREAD

**Discussion by anubra266** [remix-run/remix#1308] [Q&A] [answered]
I'm trying to create modal routes in remix.
The approach I've tried is to put my modals in a nested route, this works but will prevent closing animations which is bad for UX.

**Comment by hollandThomas** [comment]
Does it technically need to be a route? i.e. do you want to have your app's modal-state as part of your browser's history so that browser-back navigates back to an once opened modal? Do you need a deep-link that lets you link to that particular modal?

Otherwise, have a look at implementations of the [Dialog Design Pattern](https://www.w3.org/TR/wai-aria-practices-1.2/#dialog_modal) by e.g. [Reach UI](https://reach.tech/dialog) or [Radix](https://www.radix-ui.com/docs/primitives/components/dialog) and use them as-is, without creating a new nested route.

**Reply by cliffordfajardo** [reply]
> Does it technically need to be a route?

Its not necessary but there are advantages

Some pros of having a route modal
- you can share the URL and the modal will be opened and ready to use for the user
- you can use 	a `loader` and `action`  in your modal component
- as soon as my modal route is activated, remix can fetch data for the modal. For example in my modal, I have lots of `<Select>` fields and their values need to be fetched from the backend
   - If you link to the modal route in my scenario above ^^ Remix can fetch all the data in parallel on the server so it may show up faster

**Comment by cliffordfajardo** [comment]
I ran into this same issue of my route not animating on the way out.
I created a simpler codesandbox demonstrating the problem of no exit animation when the route is switched:
https://codesandbox.io/s/antd-remix--modal-no-exit-transition-forked-bd6ckh


https://user-images.githubusercontent.com/6743796/187083943-72d85cf5-389e-43e1-80e7-c061318d2c4a.mp4





Cross posting here for visibility; will update if find answer
 https://github.com/remix-run/remix/discussions/4092

**Reply by anubra266** [reply, marked as answer]
@cliffordfajardo I came up with this: 
Basically all the code for the modal is in the page page, then the outlet is in the modal code, this way, the open and close transition is handled from the parent page, you can then pass an onClose method to the modal page through context, onClose should navigate back to parent of course.
<img width="752" alt="CleanShot 2022-08-28 at 11 59 30@2x" src="https://user-images.githubusercontent.com/30869823/187070692-a5c5c4e9-686b-44b5-8f4f-2cf03512b875.png">

**Reply by jamalsoueidan** [reply]
Works thanks 👍🏻

**Reply by Roanmh** [reply]
I recently implemented this solution and have two notes from my troubleshooting:

1. I repeated this solution, but with [Shadcn/ui's Dialog](https://ui.shadcn.com/docs/components/dialog). It worked but I needed `@radix-ui/react-portal@1.1.1` (or later) to avoid a hydration error when the modal's route is loaded directly (via refresh or url entry).
2. I found that `useLocation()` also gave me the information I needed to determine if the dialog is open. The `pathname` of the location is more clear and more granular and, from what I can tell, `useOutlet()` is intended for [internal use in react-router](https://reactrouter.com/en/6.25.0/hooks/use-outlet). `useOutlet()` worked just fine though.

**Reply by aon** [reply]
I've also implemented this solution with shadcn/ui. Two notes as well to improve the solution:

1. Solving the hydration error, as pointed out by @Roanmh, is just changing the following in the `dialog.tsx` component file:

```diff
import * as DialogPrimitive from "@radix-ui/react-dialog"
+import * as Portal from "@radix-ui/react-portal"

-const DialogPortal = DialogPrimitive.Portal
+const DialogPortal = Portal.Root
```

2. In order to prevent navigating away before finishing the animation, you can make use of the `onCloseAutoFocus` prop in `DialogContent` which gets triggered after finishing the animation.

**Comment by n1ghtmare** [comment]
Hey, I wrote a [blog post ](https://n1ghtmare.github.io/2022-09-06/remix-route-modal-with-framer-motion/) about this. My version uses Framer Motion because that's what I use for animations across my project, but hey someone might find it useful.
